File: backend/app/automation/bibliography_intelligence.py
"""Deterministic bibliography compatibility engine.

No LLMs, no automatic recalibration. The engine recommends compatible evidence
and states what each source cannot support.
"""
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Literal

from pydantic import BaseModel, Field

EvidenceTag = Literal["local", "comparable", "benchmark", "solo_contexto", "no_usable"]
EvidenceType = Literal["document", "api", "benchmark", "gap", "model"]
Stage = Literal["validation", "planning", "execution"]


class BibliographyRecord(BaseModel):
    id: str
    tenant_id: str
    municipality: str
    state: str
    municipio_id: str | None = None
    inegi_clave: str | None = None
    institution: str
    title: str
    source_date: str
    consulted_at: str
    territorial_scope: Literal["municipio", "zm", "estado", "nacional"] = "municipio"
    module_id: str
    claim_id: str
    claim_label: str
    evidence_type: EvidenceType
    method: str
    confidence: str
    restrictions: list[str] = Field(default_factory=list)


class BibliographyCompatibilityScore(BaseModel):
    territorial: int
    profile: int
    module: int
    recency: int
    evidence: int
    penalties: int
    total: int


class EvidenceRecommendation(BaseModel):
    id: str
    tag: EvidenceTag
    record: BibliographyRecord
    score: BibliographyCompatibilityScore
    confidence: Literal["high", "medium", "low", "blocked"]
    supported_claim: str
    unsupported_claim: str
    explanation: str
    stage: Stage
    module_id: str


class StageEvidenceMap(BaseModel):
    stage: Stage
    label: str
    recommendations: list[EvidenceRecommendation]
    local_count: int
    comparable_count: int
    benchmark_count: int
    blocked_count: int


STAGE_MODULES: dict[Stage, list[str]] = {
    "validation": ["M00B", "M01", "M02", "M03B", "M04", "M13", "M14", "M15"],
    "planning": ["M05", "M07", "M08", "M09", "M13", "M14"],
    "execution": ["M17", "M18", "M20", "M21"],
}
STAGE_LABELS = {"validation": "Validacion", "planning": "Planeacion", "execution": "Ejecucion"}

FIELD_CONTRACTS = {
    "antecedentes.demografia.poblacion": ("M01", "Poblacion municipal"),
    "antecedentes.demografia.viviendas": ("M01", "Viviendas municipales"),
    "antecedentes.demografia.generacion_kg_hab_dia": ("M01", "Generacion kg/hab/dia"),
}


def _now_date() -> str:
    return datetime.now(timezone.utc).date().isoformat()


def _unwrap_field(profile: dict[str, Any], dotted: str) -> dict[str, Any] | None:
    current: Any = profile
    for part in dotted.split("."):
        if not isinstance(current, dict):
            return None
        current = current.get(part)
    return current if isinstance(current, dict) else None


def _evidence_type(field: dict[str, Any]) -> EvidenceType:
    if field.get("value") is None:
        return "gap"
    method = str(field.get("method") or "").lower()
    source = field.get("source") or {}
    label = str(source.get("label") or "").lower()
    if "benchmark" in method or "benchmark" in label:
        return "benchmark"
    if "api" in method:
        return "api"
    if "inferred" in str(field.get("confidence") or ""):
        return "model"
    return "document"


def _territorial_scope(field: dict[str, Any]) -> Literal["municipio", "zm", "estado", "nacional"]:
    method = str(field.get("method") or "").lower()
    source = field.get("source") or {}
    label = str(source.get("label") or "").lower()
    if "benchmark" in method or "semarnat" in label:
        return "nacional"
    return "municipio"


def _restrictions(scope: str, evidence_type: EvidenceType) -> list[str]:
    restrictions: list[str] = []
    if scope != "municipio":
        restrictions.append("No soporta verdad municipal directa.")
    if scope == "zm":
        restrictions.append("Alcance ZM; debe separarse de municipio.")
    if evidence_type == "benchmark":
        restrictions.append("Benchmark; no sustituye estudio local.")
    if evidence_type == "gap":
        restrictions.append("Brecha; no usable como afirmacion.")
    if evidence_type == "model":
        restrictions.append("Inferencia/modelo; requiere revision humana.")
    return restrictions


def build_bibliography_records(tenants: list[dict[str, Any]]) -> list[BibliographyRecord]:
    records: list[BibliographyRecord] = []
    for tenant in tenants:
        profile = tenant.get("municipal_profile") or {}
        for field_path, (module_id, label) in FIELD_CONTRACTS.items():
            field = _unwrap_field(profile, field_path)
            if not field:
                continue
            source = field.get("source") or {}
            evidence_type = _evidence_type(field)
            scope = _territorial_scope(field)
            source_date = source.get("extracted_at") or tenant.get("updated_at") or _now_date()
            records.append(BibliographyRecord(
                id=f"{tenant.get('id')}:{field_path}",
                tenant_id=str(tenant.get("id")),
                municipality=str(tenant.get("nombre")),
                state=str(tenant.get("estado_mx")),
                municipio_id=tenant.get("municipio_id"),
                inegi_clave=tenant.get("inegi_clave"),
                institution=str(source.get("label") or "Fuente pendiente"),
                title=label,
                source_date=str(source_date),
                consulted_at=str(source_date),
                territorial_scope=scope,
                module_id=module_id,
                claim_id=field_path,
                claim_label=label,
                evidence_type=evidence_type,
                method=str(field.get("method") or "metodo_pendiente"),
                confidence=str(field.get("confidence") or "pending_human_validation"),
                restrictions=_restrictions(scope, evidence_type),
            ))
    return records


def _year(value: str) -> int | None:
    import re
    match = re.search(r"20\d{2}|19\d{2}", value)
    return int(match.group(0)) if match else None


def _recency(value: str) -> int:
    year = _year(value)
    if not year:
        return 8
    age = max(0, 2026 - year)
    if age <= 2:
        return 15
    if age <= 5:
        return 10
    if age <= 10:
        return 5
    return 0


def _confidence_score(confidence: str) -> int:
    if "high" in confidence or "official" in confidence:
        return 18
    if "medium" in confidence or "secondary" in confidence:
        return 10
    if "low" in confidence or "pending" in confidence:
        return 4
    if "critical" in confidence:
        return -40
    return 6


def _stage_for(module_id: str) -> Stage:
    if module_id in STAGE_MODULES["execution"]:
        return "execution"
    if module_id in STAGE_MODULES["planning"]:
        return "planning"
    return "validation"


def score_record(target: dict[str, Any], record: BibliographyRecord, module_id: str | None = None) -> BibliographyCompatibilityScore:
    same_municipality = bool(target.get("municipio_id") and target.get("municipio_id") == record.municipio_id)
    same_state = target.get("estado_mx") == record.state
    territorial = 45 if same_municipality and record.territorial_scope == "municipio" else 16 if same_state else 10 if record.territorial_scope == "nacional" else 8
    profile = 12 if same_state else 6
    module = 10 if not module_id else 18 if module_id == record.module_id else 0
    recency = _recency(record.source_date)
    evidence = _confidence_score(record.confidence)
    penalties = sum([
        -80 if record.evidence_type == "gap" else 0,
        -12 if record.evidence_type == "benchmark" else 0,
        -10 if record.evidence_type == "model" else 0,
        -8 if record.territorial_scope != "municipio" else 0,
    ])
    total = max(0, min(100, territorial + profile + module + recency + evidence + penalties))
    return BibliographyCompatibilityScore(
        territorial=territorial,
        profile=profile,
        module=module,
        recency=recency,
        evidence=evidence,
        penalties=penalties,
        total=total,
    )


def _tag(target: dict[str, Any], record: BibliographyRecord, score: BibliographyCompatibilityScore) -> EvidenceTag:
    if record.evidence_type == "gap" or score.total < 20:
        return "no_usable"
    if record.territorial_scope == "municipio" and record.municipio_id == target.get("municipio_id") and record.evidence_type == "document":
        return "local"
    if record.evidence_type == "benchmark" or record.territorial_scope == "nacional":
        return "benchmark"
    if record.territorial_scope in {"zm", "estado"}:
        return "comparable"
    return "comparable" if score.total >= 45 else "solo_contexto"


def _recommendation_confidence(tag: EvidenceTag, score: int) -> Literal["high", "medium", "low", "blocked"]:
    if tag == "no_usable":
        return "blocked"
    if tag == "local" and score >= 70:
        return "high"
    if score >= 50:
        return "medium"
    return "low"


def build_evidence_recommendations(
    target: dict[str, Any],
    tenants: list[dict[str, Any]],
    *,
    stage: Stage | None = None,
    module_id: str | None = None,
) -> list[EvidenceRecommendation]:
    records = build_bibliography_records(tenants)
    if stage:
        records = [record for record in records if record.module_id in STAGE_MODULES[stage]]
    if module_id:
        records = [record for record in records if record.module_id == module_id]
    recommendations: list[EvidenceRecommendation] = []
    for record in records:
        score = score_record(target, record, module_id)
        tag = _tag(target, record, score)
        recommendations.append(EvidenceRecommendation(
            id=f"{target.get('id')}:{record.id}",
            tag=tag,
            record=record,
            score=score,
            confidence=_recommendation_confidence(tag, score.total),
            supported_claim=(
                f"Puede soportar claim municipal sobre {record.claim_label}."
                if tag == "local"
                else "No debe soportar afirmaciones." if tag == "no_usable"
                else f"Puede contextualizar {record.claim_label} como evidencia comparable o benchmark."
            ),
            unsupported_claim=(
                "No sustituye revision humana ni oficialidad automatica."
                if tag == "local"
                else "No soporta verdad municipal, estudio local ni declaratoria oficial."
            ),
            explanation=(
                f"Fuente municipal de {record.municipality}; usable con fuente, fecha, metodo, alcance y revision humana."
                if tag == "local"
                else "La plataforma recomienda esta fuente como contexto comparable; no es estudio local ni dato oficial municipal."
            ),
            stage=_stage_for(record.module_id),
            module_id=record.module_id,
        ))
    return sorted(recommendations, key=lambda item: item.score.total, reverse=True)


def build_stage_evidence_map(target: dict[str, Any], tenants: list[dict[str, Any]]) -> list[StageEvidenceMap]:
    output: list[StageEvidenceMap] = []
    for stage in ("validation", "planning", "execution"):
        recommendations = build_evidence_recommendations(target, tenants, stage=stage)[:8]
        output.append(StageEvidenceMap(
            stage=stage,  # type: ignore[arg-type]
            label=STAGE_LABELS[stage],  # type: ignore[index]
            recommendations=recommendations,
            local_count=sum(1 for item in recommendations if item.tag == "local"),
            comparable_count=sum(1 for item in recommendations if item.tag == "comparable"),
            benchmark_count=sum(1 for item in recommendations if item.tag == "benchmark"),
            blocked_count=sum(1 for item in recommendations if item.tag == "no_usable"),
        ))
    return output

File: frontend/src/lib/bibliographyIntelligence.ts
import type { CitationRecord, TenantDiagnosticData, TenantMetric } from '@/lib/tenantDiagnosticData'
import { CITATION_REGISTRY } from '@/lib/tenantDiagnosticData'

export type EvidenceRecommendationTag = 'local' | 'comparable' | 'benchmark' | 'solo_contexto' | 'no_usable'
export type BibliographyEvidenceType = 'document' | 'api' | 'benchmark' | 'gap' | 'model'
export type PlatformStage = 'validation' | 'planning' | 'execution'

export interface BibliographyRecord {
  id: string
  tenant_id: string
  municipality: string
  state: string
  municipio_id?: string
  zm?: string
  citation_id: string
  institution: string
  title: string
  parent_document?: string
  url?: string
  source_date: string
  consulted_at: string
  territorial_scope: TenantMetric['territorial_scope']
  module_id: string
  claim_id: string
  claim_label: string
  evidence_type: BibliographyEvidenceType
  method: string
  confidence: TenantMetric['confidence']
  restrictions: string[]
}

export interface BibliographyCompatibilityScore {
  territorial: number
  profile: number
  module: number
  recency: number
  evidence: number
  penalties: number
  total: number
}

export interface EvidenceRecommendation {
  id: string
  tag: EvidenceRecommendationTag
  record: BibliographyRecord
  score: BibliographyCompatibilityScore
  confidence: 'high' | 'medium' | 'low' | 'blocked'
  supported_claim: string
  unsupported_claim: string
  explanation: string
  stage: PlatformStage
  module_id: string
}

export interface StageEvidenceMap {
  stage: PlatformStage
  label: string
  recommendations: EvidenceRecommendation[]
  local_count: number
  comparable_count: number
  benchmark_count: number
  blocked_count: number
}

const STAGE_MODULES: Record<PlatformStage, string[]> = {
  validation: ['M00B', 'M01', 'M02', 'M03B', 'M04', 'M13', 'M14', 'M15'],
  planning: ['M05', 'M07', 'M08', 'M09', 'M13', 'M14'],
  execution: ['M17', 'M18', 'M20', 'M21'],
}

const STAGE_LABELS: Record<PlatformStage, string> = {
  validation: 'Validación',
  planning: 'Planeación',
  execution: 'Ejecución',
}

const METRIC_MODULE: Record<string, string> = {
  rsu_generation: 'M01',
  field_characterization: 'M01',
  collection_coverage: 'M08',
  routes_time_study: 'M08',
  psp_acceptance: 'M13',
}

function moduleForMetric(metric: TenantMetric): string {
  if (metric.field_id?.startsWith('M')) return metric.field_id.split('.')[0]
  return METRIC_MODULE[metric.id] ?? 'M00B'
}

function evidenceType(metric: TenantMetric, citation: CitationRecord): BibliographyEvidenceType {
  if (metric.status === 'brecha_critica' || metric.confidence === 'critical_gap') return 'gap'
  if (metric.status === 'inferido') return 'model'
  const text = `${citation.institution} ${citation.title} ${citation.parent_document ?? ''}`.toLowerCase()
  if (metric.territorial_scope === 'nacional' || text.includes('benchmark')) return 'benchmark'
  if (metric.source.toLowerCase().includes('api')) return 'api'
  return 'document'
}

function yearFromDate(value: string): number | null {
  const match = value.match(/20\d{2}|19\d{2}/)
  return match ? Number(match[0]) : null
}

function recencyScore(value: string) {
  const year = yearFromDate(value)
  if (!year) return 8
  const age = Math.max(0, 2026 - year)
  if (age <= 2) return 15
  if (age <= 5) return 10
  if (age <= 10) return 5
  return 0
}

function confidenceScore(confidence: TenantMetric['confidence']) {
  if (confidence === 'verified_official') return 20
  if (confidence === 'verified_secondary') return 16
  if (confidence === 'inferred_medium') return 8
  if (confidence === 'inferred_low' || confidence === 'pending_validation') return 4
  return -40
}

function recommendationStage(moduleId: string): PlatformStage {
  if (STAGE_MODULES.execution.includes(moduleId)) return 'execution'
  if (STAGE_MODULES.planning.includes(moduleId)) return 'planning'
  return 'validation'
}

export function buildBibliographyRecords(tenants: TenantDiagnosticData[]): BibliographyRecord[] {
  return tenants.flatMap(tenant =>
    tenant.metrics.flatMap(metric => {
      if (!metric.citation_id) return []
      const citation = CITATION_REGISTRY[metric.citation_id]
      if (!citation) return []
      const moduleId = moduleForMetric(metric)
      const type = evidenceType(metric, citation)
      return [{
        id: `${tenant.tenant_id}:${metric.id}:${citation.id}`,
        tenant_id: tenant.tenant_id,
        municipality: tenant.municipality,
        state: tenant.state,
        municipio_id: tenant.municipio_id,
        zm: tenant.zm,
        citation_id: citation.id,
        institution: citation.institution,
        title: citation.title,
        parent_document: citation.parent_document,
        url: citation.url,
        source_date: metric.source_date || citation.year_or_date,
        consulted_at: metric.consulted_at || citation.consulted_at,
        territorial_scope: metric.territorial_scope,
        module_id: moduleId,
        claim_id: metric.id,
        claim_label: metric.label,
        evidence_type: type,
        method: metric.method,
        confidence: metric.confidence,
        restrictions: restrictionsFor(metric, type),
      }]
    }),
  )
}

function restrictionsFor(metric: TenantMetric, type: BibliographyEvidenceType): string[] {
  const restrictions: string[] = []
  if (metric.territorial_scope !== 'municipio') restrictions.push('No soporta verdad municipal directa.')
  if (metric.territorial_scope === 'zm') restrictions.push('Alcance ZM; debe separarse de municipio.')
  if (type === 'benchmark') restrictions.push('Benchmark; no sustituye estudio local.')
  if (type === 'gap') restrictions.push('Brecha; no usable como afirmación.')
  if (type === 'model') restrictions.push('Inferencia/modelo; requiere revisión humana.')
  return restrictions
}

function scoreRecord(target: TenantDiagnosticData, record: BibliographyRecord, moduleId?: string): BibliographyCompatibilityScore {
  const sameMunicipality = Boolean(target.municipio_id && record.municipio_id === target.municipio_id)
  const sameState = target.state === record.state
  const sameZm = Boolean(target.zm && record.zm === target.zm)
  const territorial = sameMunicipality && record.territorial_scope === 'municipio'
    ? 45
    : sameZm && record.territorial_scope === 'zm'
      ? 22
      : sameState
        ? 16
        : record.territorial_scope === 'nacional'
          ? 10
          : 8
  const profile = sameState ? 12 : 6
  const moduleScore = !moduleId ? 10 : record.module_id === moduleId ? 18 : 0
  const recency = recencyScore(record.source_date)
  const evidence = confidenceScore(record.confidence)
  const penalties = [
    record.evidence_type === 'gap' ? -80 : 0,
    record.evidence_type === 'benchmark' ? -12 : 0,
    record.evidence_type === 'model' ? -10 : 0,
    record.territorial_scope === 'zm' && !sameMunicipality ? -15 : 0,
    record.territorial_scope !== 'municipio' ? -8 : 0,
  ].reduce((sum, item) => sum + item, 0)
  const total = Math.max(0, Math.min(100, territorial + profile + moduleScore + recency + evidence + penalties))
  return { territorial, profile, module: moduleScore, recency, evidence, penalties, total }
}

function tagFor(target: TenantDiagnosticData, record: BibliographyRecord, score: BibliographyCompatibilityScore): EvidenceRecommendationTag {
  if (record.evidence_type === 'gap' || score.total < 20) return 'no_usable'
  if (record.territorial_scope === 'municipio' && target.municipio_id && record.municipio_id === target.municipio_id && record.evidence_type === 'document') return 'local'
  if (record.evidence_type === 'benchmark' || record.territorial_scope === 'nacional') return 'benchmark'
  if (record.territorial_scope === 'zm' || record.territorial_scope === 'estado') return 'comparable'
  return score.total >= 45 ? 'comparable' : 'solo_contexto'
}

function recommendationConfidence(tag: EvidenceRecommendationTag, score: number): EvidenceRecommendation['confidence'] {
  if (tag === 'no_usable') return 'blocked'
  if (tag === 'local' && score >= 70) return 'high'
  if (score >= 50) return 'medium'
  return 'low'
}

export function buildEvidenceRecommendations(
  target: TenantDiagnosticData,
  allTenants: TenantDiagnosticData[],
  options: { stage?: PlatformStage; module_id?: string } = {},
): EvidenceRecommendation[] {
  const stageModules = options.stage ? STAGE_MODULES[options.stage] : null
  return buildBibliographyRecords(allTenants)
    .filter(record => !stageModules || stageModules.includes(record.module_id))
    .filter(record => !options.module_id || record.module_id === options.module_id)
    .map(record => {
      const score = scoreRecord(target, record, options.module_id)
      const tag = tagFor(target, record, score)
      const stage = recommendationStage(record.module_id)
      return {
        id: `${target.tenant_id}:${record.id}`,
        tag,
        record,
        score,
        confidence: recommendationConfidence(tag, score.total),
        supported_claim: tag === 'local'
          ? `Puede soportar claim municipal sobre ${record.claim_label}.`
          : tag === 'no_usable'
            ? 'No debe soportar afirmaciones.'
            : `Puede contextualizar ${record.claim_label} como evidencia comparable o benchmark.`,
        unsupported_claim: tag === 'local'
          ? 'No sustituye revisión humana ni oficialidad automática.'
          : 'No soporta verdad municipal, estudio local ni declaratoria oficial.',
        explanation: explanationFor(target, record, tag),
        stage,
        module_id: record.module_id,
      }
    })
    .sort((a, b) => b.score.total - a.score.total)
}

function explanationFor(target: TenantDiagnosticData, record: BibliographyRecord, tag: EvidenceRecommendationTag): string {
  if (tag === 'local') {
    return `Fuente municipal de ${record.municipality}; puede alimentar claims si conserva fuente, fecha, método, alcance y revisión humana.`
  }
  if (tag === 'benchmark') {
    return 'La plataforma recomienda esta fuente como contexto comparable; no es estudio local ni dato oficial municipal.'
  }
  if (tag === 'comparable') {
    return `Fuente comparable para ${target.municipality}; útil para hipótesis, riesgos o contexto, no para cerrar verdad municipal.`
  }
  if (tag === 'solo_contexto') {
    return 'Fuente de bajo ajuste; usar sólo como contexto metodológico.'
  }
  return 'Fuente no usable para afirmaciones por brecha, baja confianza o incompatibilidad territorial.'
}

export function buildStageEvidenceMap(target: TenantDiagnosticData, tenants: TenantDiagnosticData[]): StageEvidenceMap[] {
  return (['validation', 'planning', 'execution'] as PlatformStage[]).map(stage => {
    const recommendations = buildEvidenceRecommendations(target, tenants, { stage }).slice(0, 8)
    return {
      stage,
      label: STAGE_LABELS[stage],
      recommendations,
      local_count: recommendations.filter(item => item.tag === 'local').length,
      comparable_count: recommendations.filter(item => item.tag === 'comparable').length,
      benchmark_count: recommendations.filter(item => item.tag === 'benchmark').length,
      blocked_count: recommendations.filter(item => item.tag === 'no_usable').length,
    }
  })
}
